File: benchmark-java-sync/src/main/java/com/example/benchmarkjavasync/api/ServiceAdapter.java
package com.example.benchmarkjavasync.api;

import com.example.benchmarkjavasync.core.ApplicationService;
import com.example.benchmarkjavasync.core.model.ApplicationResponse;
import com.example.benchmarkjavasync.core.model.ApplicationServiceRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ServiceAdapter {

    private final ApplicationService<Optional<ApplicationResponse>> applicationService;

    public Optional<ApplicationResponse> process(ApplicationRequestDto applicationRequestDto) {
        return applicationService.process(map(applicationRequestDto));

    }

    public ApplicationServiceRequest map(ApplicationRequestDto applicationRequestDto) {
        return new ApplicationServiceRequest(
                applicationRequestDto.firstName(),
                applicationRequestDto.lastName(),
                applicationRequestDto.passportSerialNumber(),
                applicationRequestDto.passportNumber(),
                applicationRequestDto.amount(),
                applicationRequestDto.status()
        );
    }
}

File: benchmark-kotlin-async/src/main/kotlin/com/example/benchmarkkotlinasync/api/ServiceAdapter.kt
package com.example.benchmarkkotlinasync.api;

import com.example.benchmarkkotlinasync.core.ApplicationService
import com.example.benchmarkkotlinasync.core.model.ApplicationResponse
import com.example.benchmarkkotlinasync.core.model.ApplicationServiceRequest
import org.springframework.stereotype.Service

@Service
class ServiceAdapter(private val applicationService: ApplicationService<ApplicationResponse>) {

    suspend fun process(applicationRequestDto: ApplicationRequestDto): ApplicationResponse =
        applicationService.process(map(applicationRequestDto))

    suspend fun map(applicationRequestDto: ApplicationRequestDto): ApplicationServiceRequest =
        ApplicationServiceRequest(
            applicationRequestDto.firstName,
            applicationRequestDto.lastName,
            applicationRequestDto.passportSerialNumber,
            applicationRequestDto.passportNumber,
            applicationRequestDto.amount,
            applicationRequestDto.status
        )
}
